main.lua

```diff
@@ -957,9 +959,52 @@ function onCursorRight(bufpane)    clearAutocomplete() end
 function onCursorStart(bufpane)    clearAutocomplete() end
 function onCursorEnd(bufpane)      clearAutocomplete() end
 
-function onUndo(bp) fullyUpdate(bp.Buf) end
-function onRedo(bp) fullyUpdate(bp.Buf) end
+function preUndo(bp)
+    undoStackLengthBefore = bp.Buf.UndoStack:Len()
+end
+function onUndo(bp)
+    local numUndos = undoStackLengthBefore - bp.Buf.UndoStack:Len()
+    return handleUndosRedos(bp.Buf, bp.Buf.RedoStack.Top, numUndos)
+end
+
+function preRedo(bp)
+    undoStackLengthBefore = bp.Buf.UndoStack:Len()
+end
+function onRedo(bp)
+    local numRedos = bp.Buf.UndoStack:Len() - undoStackLengthBefore
```

Comment:
Same here, wrong stack. The "snapshot" should be taken for the stack of the action. And the elements need to be extracted from the opposite.

preSTACK -> get the length of STACK
onSTACK -> calculate the change + check the elements in **opposite** STACK

Reply:
And should be like the Undo one, `stackBefore - stack:Len() `. In `pre` always should be more than in `on`, it's a LIFO.

Reply:
It's intentionally using the same stack for both Redo/Undo – redoing moves items from `RedoStack` to `UndoStack` so the `UndoStack` gets *longer* (as opposed to getting shorter when undoing), hence the calculation is the other way around in `onRedo` compared to `onUndo`.

Reply:
You are right, this is other way. See thats the tricky happening hahah, the other implementation has only one way.